src/cockpit/channels/metrics.py

```diff
@@ -15,23 +15,192 @@
+import time
 import logging
+from typing import Any, Dict, List, NamedTuple, Optional
 
-from ..channel import Channel
+from ..channel import AsyncChannel
+from .samples import Sampler, SAMPLE_DESCRIPTIONS
+from .samples import (cockpit_cpu_samples, cockpit_memory_samples, cockpit_cpu_temperature_samples)
 
 logger = logging.getLogger(__name__)
 
 
-class MetricsChannel(Channel):
+class Instance(NamedTuple):
+    old_value: Any
+    value: Any
+    seen: bool
+
+
+class MetricDescription(NamedTuple):
+    name: str
+    units: str
+    semantics: str
+    instanced: bool
+    sampler: int
+
+
+class MetricInfo(NamedTuple):
+    desc: MetricDescription
+    derive: Optional[str]
+    instances: Dict[str, Instance]
+    value: Any
+    old_value: Any
+
+
+class InternalMetricsChannel(AsyncChannel):
     payload = 'metrics1'
+    restrictions = {'source': 'internal'}
+    metrics: List[MetricInfo]
+    samplers: int = 0
+    interval: int = 1000
+    # If any metric "subscriptions" change, we need to inform the client.
+    need_meta: bool = False
+    last_timestamp: int = 0
+    next_timestamp: int = 0
+
+    def parse_options(self, options):
+        logger.debug("metrics internal open: %s, channel: %s", options, self.channel)
+        interval = options.get('interval', self.interval)
+        if interval <= 0 or interval > sys.maxsize:
+            logger.error('invalid "interval" value: %s', interval)
+            self.close(problem="protocol-error", message='invalid interval option')
+            return False
+
+        self.interval = interval
+
+        metrics = options.get('metrics')
+        if not metrics or not isinstance(metrics, list):
+            logger.error('invalid "metrics" value: %s', metrics)
+            self.close(problem='metrics is not an list')
+            return False
+
+        for metric in metrics:
+            # validate it's an object
+            name = metric.get('name')
+            units = metric.get('units')
+            derive = metric.get('derive')
+
+            desc = SAMPLE_DESCRIPTIONS.get(name)
+            if not desc:
+                logger.error('unsupported metric: %s', name)
+                self.close(problem='not-supported',
+                           message='invalid metric')
+                return False
+
+            (unit, semantics, instanced, sampler) = desc
+            if units and units != unit:
+                self.close(problem='protocol-error',
+                           message=f'{name} has units {unit}, not {units}')
+                return False
+
+            # Keep track of what samplers are requested
+            self.samplers |= sampler.value
+
+            desc = MetricDescription(name, unit, semantics, instanced, sampler.value)
+            self.metrics.append(MetricInfo(desc, derive, {}, None, None))
+
+        return True
+
+    def cockpit_internal_metrics_sample(self, metric, instance=None, value=None):
+        for index, metricinfo in enumerate(self.metrics):
```

Comment:
This loop seems inefficient.  I think you want some kind of a lookup table here, no?

Reply:
Lookup sounds like the way to go yes, I had this on my mind, forgot...